### DISCUSSION TITLE
How can I add a section to where the meta (title/author/date) information is placed in HTML documents?

### DISCUSSION THREAD

**Discussion by coatless** [quarto-dev/quarto-cli#6475] [Q&A]
### Description

I'm interested in being able to add values associated with a custom key into the document's header section through a quarto filter extension. The header section I'm interested in is where the title/subtitle/author/date information may be found, _not_ the documents `<head></head>` section. I'm specifically hoping for something along the lines of:

```
quarto.doc.include_meta_data("custom key", "custom value")
```

or

```
quarto.doc.include_meta_html("console.html")
```

For example, with an HTML document, I would like to add to the `quarto-title-meta` class div:

<img width="534" alt="quarto-annotated-key" src="https://github.com/quarto-dev/quarto-cli/assets/833642/0dddcfb5-c499-4e0e-b886-93e8ecc0ef9c">

The output would be:

```html
<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
....
</div>
  <div class="quarto-title-meta">
 ...
   <div>
    <div class="quarto-title-meta-heading">Custom Key</div>
    <div class="quarto-title-meta-contents">
      <p>Custom Value</p>
    </div>
   </div>
  </div>
</header>
```

**Note:** If either the `author` or `date` key is missing, then the `quarto-title-meta` div is not incorporated into the output. 
 
It would also be great if an HTML file could be incorporated without a quarto-specific structure. For example, I would like to incorporate an "expandable" console with a button action free from the above quarto classes:

<img width="369" alt="quarto-custom-html-annotated" src="https://github.com/quarto-dev/quarto-cli/assets/833642/57944928-5e80-4ff4-9b5b-954e352ddb66">

That said, I'm looking for the hook to work in a similar manner to an HTML book + website:

<img width="587" alt="quarto-book-html-annotated" src="https://github.com/quarto-dev/quarto-cli/assets/833642/ecc4a7fd-123c-4253-af15-46a214559a2e">

<img width="434" alt="quarto-website-annotation" src="https://github.com/quarto-dev/quarto-cli/assets/833642/01acd700-c088-48eb-9a2d-00e9680a59df">

**Comment by mcanouil** [comment]
The following source files might help you as the header is not completely made with a template, it uses TypeScript:

- https://github.com/quarto-dev/quarto-cli/blob/174828920f27decc77f627682d3404da494ca146/src/format/html/format-html-bootstrap.ts#L576

On the other hand, the new Manuscript project type uses an HTML template that could get inspiration from as a partial rather than from within a Lua filter:

- https://github.com/quarto-dev/quarto-cli/blob/174828920f27decc77f627682d3404da494ca146/src/resources/formats/html/templates/manuscript/title-metadata.html

**Reply by coatless** [reply]
So, I think the suggestion here is to ship a new [HTML partial `title-metadata.html` template](https://github.com/quarto-dev/quarto-cli/tree/main/src/resources/formats/html/pandoc) based on the content in the [Article Templates](https://quarto.org/docs/journals/templates.html#html-partials) help documentation? I think this indicates that variables set by a Lua filter be available to pandoc so that we could use `$extension.customkey$`?

Is there a good Quarto documentation page on using partial templates for HTML? (Not LaTeX)

From the LaTeX documentation, I'm thinking the extension structure likely needs to be like so:

```
ghusername/extendme
└── _extensions
    └── extendme
        ├── _extension.yaml
        ├── extend.lua
        └── partials
           ├── title-slide.html
           └── title-metadata.html
```

Then, we would need to override both the `html` and `revealjs` formats? (Is there a way to supply a book/web template?)

```yaml
format:
  html:
    template-partials:
      - title-metadata.html
  revealjs:
    template-partials:
      - title-slide.html
```

**Reply by mcanouil** [reply]
There is the following general page on partial with a brief section (linking to the source code) on HTML partial, see <https://quarto.org/docs/journals/templates.html#html-partials>.

**Reply by coatless** [reply]
Linked there ;) So, I'm guessing I get to explore partials a bit more this weekend. :)

**Reply by cderv** [reply]
Yes we don't have more extensive documentation on HTML partials, but your suggestion with extensions seems ok. 

> It would also be great if an HTML file could be incorporated without a quarto-specific structure. For example, I would like to incorporate an "expandable" console with a button action free from the above quarto classes:

This sounds like a possible feature request on its own maybe to have this part of default more HTML customizable maybe 🤔

**Comment by the-solipsist** [comment]
I'd like to be able to do this as well. I'm trying to archive all my writings on a single website. While doing so, I'd like to also add the original `publication` and the `url` as custom fields to the title block. I thought adding this as a `title-metadata.html` template-partial would help:

```
$if(by-affiliation/first)$
<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">$labels.authors$</div>
  <div class="quarto-title-meta-heading">$labels.affiliations$</div>
  
  $for(by-author)$
  <div class="quarto-title-meta-contents">
    <p class="author">$_title-meta-author.html()$</p>
  </div>
  $if(by-author.affiliations)$
  <div class="quarto-title-meta-contents">
    $for(by-author.affiliations)$
    <p class="affiliation">
      $if(it.url)$
      <a href="$it.url$">
      $endif$
      $it.name$
      $if(it.url)$
      </a>
      $endif$
    </p>
    $endfor$
  </div>
  $endif$
  $endfor$
</div>
$endif$

<div class="quarto-title-meta">

  $if(by-affiliation)$
  $elseif(by-author)$
  <div>
    <div class="quarto-title-meta-heading">$labels.authors$</div>
       <p>$_title-meta-author.html()$</p>
    <div class="quarto-title-meta-contents">
      $for(by-author)$
      $endfor$
    </div>
  </div>
  $endif$
  
  $if(date)$
  <div>
    <div class="quarto-title-meta-heading">$labels.published$</div>
    <div class="quarto-title-meta-contents">
      <p class="date">$date$</p>
    </div>
  </div>
  $endif$

  $if(date-modified)$
  <div>
    <div class="quarto-title-meta-heading">$labels.modified$</div>
    <div class="quarto-title-meta-contents">
      <p class="date-modified">$date-modified$</p>
    </div>
  </div>
  $endif$
  
  $if(doi)$
  <div>
    <div class="quarto-title-meta-heading">$labels.doi$</div>
    <div class="quarto-title-meta-contents">
      <p class="doi">
        <a href="https://doi.org/$doi$">$doi$</a>
      </p>
    </div>
  </div>
  $endif$

  $if(publication)$
  <div>
    <div class="quarto-title-meta-heading">Publication</div>
    <div class="quarto-title-meta-contents">
      <p class="publication">$publication$</p>
    </div>
  </div>
  $endif$
  
  $if(url)$
  <div>
    <div class="quarto-title-meta-heading">URL</div>
    <div class="quarto-title-meta-contents">
      <p class="url">
        <a href="$url$">$url$</a>
      </p>
    </div>
  </div>
  $endif$
  $if(archive)$
  <div>
    <div class="quarto-title-meta-heading">Archive</div>
    <div class="quarto-title-meta-contents">
      <p class="url">
        <a href="$archive$">$archive$</a>
      </p>
    </div>
  </div>
  $endif$
</div>

$if(abstract)$
<div>
  <div class="abstract">
    <div class="block-title">$labels.abstract$</div>
    $abstract$
  </div>
</div>
$endif$

$if(keywords)$
<div>
  <div class="keywords">
    <div class="block-title">$labels.keywords$</div>
    <p>$for(keywords)$$it$$sep$, $endfor$</p>
  </div>
</div>
$endif$
```

But if I add it with the two instances of `$_title-meta-author.html()$`, then the document won't render.  I get the following error:
`Could not find data file templates/_title-meta-author.html`.  If I remove those two instances, then the document renders, but without the author name.

**Reply by mcanouil** [reply]
You need to provide the missing file if you replace the whole template or part of it.
Take a look at the partials, see <https://quarto.org/docs/journals/templates.html>.

**Reply by the-solipsist** [reply]
I'd already tried adding `_title-meta-author.html` and `template/_title-meta-author.html` (with the content of [`_title-meta-author.html`](https://github.com/quarto-dev/quarto-cli/blob/9f379a2a7af9d9fb156f21b77f34aff23cc5a115/src/resources/formats/html/templates/_title-meta-author.html)), but to no avail.

Maybe I'm misunderstanding template-partials.  Can I add just the following without adding a whole `title-metadata.html`?

```
  $if(publication)$
  <div>
    <div class="quarto-title-meta-heading">Publication</div>
    <div class="quarto-title-meta-contents">
      <p class="publication">$publication$</p>
    </div>
  </div>
  $endif$
  
  $if(url)$
  <div>
    <div class="quarto-title-meta-heading">URL</div>
    <div class="quarto-title-meta-contents">
      <p class="url">
        <a href="$url$">$url$</a>
      </p>
    </div>
  </div>
  $endif$

  $if(archive)$
  <div>
    <div class="quarto-title-meta-heading">Archive</div>
    <div class="quarto-title-meta-contents">
      <p class="url">
        <a href="$archive$">$archive$</a>
      </p>
    </div>
  </div>
  $endif$
```

**Reply by the-solipsist** [reply]
Just wanted to check again: have I misunderstood template partials? Or have I run into a bug of some kind?

**Reply by cderv** [reply]
How are you passing partials to your document ? 

You may want to open a new discussion with a small repo example that reproduce what is not working for you. It would be easier to have help
